File: Examples/PY/test8.py
from collections.abc import Callable
def map[A, B](f: Callable[[A], B], l: list[A]) -> list[B]:
    def map2(acc: list[B], l: list[A]) -> list[B]:
        match l:
            case [hd, *tl]:
                return map2([f(hd), *acc], tl)
            case _:
                return acc
    return map2([], l)

File: Examples/TS/test8.ts
import { Callable } from "./collections.abc";
function map<A, B>(f: (arg0: A) => B, l: A[]): B[] {
    function map2(acc: B[], l: A[]): B[] {
        if (l.length === 0) {
            return acc;
        } else {
            const hd= l[0];
            const tl= l.slice(1);
            return map2([f(hd), ...acc], tl);
        }
    }
    return map2([], l);
}
